hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/InvocationMonitor.java

```diff
@@ -151,10 +153,21 @@ private void scan() {
             long now = Clock.currentTimeMillis();
             int backupTimeouts = 0;
             int invocationTimeouts = 0;
-            for (Invocation invocation : invocationRegistry.invocations()) {
+
+            Set<Map.Entry<Long, Invocation>> invocations = invocationRegistry.entrySet();
+            Iterator<Map.Entry<Long, Invocation>> iterator = invocations.iterator();
+            while (iterator.hasNext()) {
                 if (shutdown) {
                     return;
                 }
+                Map.Entry<Long, Invocation> entry = iterator.next();
+                Long callId = entry.getKey();
+                Invocation invocation = entry.getValue();
+
+                if (!callIdMatches(callId, invocation) && isDone(invocation)) {
```

Comment:
The 'phantom' invocation is removed. So if you get 2 concurrent retries the same invocation is registered under 2 different calld ids' (first retry callid1 and second retry calldi2) So you can potentially end up with phantom invocations in your registry.